### PR TITLE
[MO] [Kaldi] Add TDNN Component

### PR DESCRIPTION
**Pull request by pavel-esir**
Description: added to MO support TDNN Component from Kaldi. TDNN component is expanded into operations from OpenVINO opset.

#35123, #27253

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR: N/A -- original component is not reshape-able
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests: manual check showed that infer results match with the reference
* [x]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list
* [x]  Supported **public** models list: N/A
* [x]  New operations specification: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

### CODE REVIEW

File: model-optimizer/mo/ops/memoryoffset.py [hunk 1]
```diff
@@ -58,9 +57,19 @@ def infer(node: Node):
...
+            elif pair_node.in_port(0).get_source().node.op in ["Add", "ReLU"]:
+                if pair_node.in_port(0).get_source().node.in_port(0).get_source().node.has_valid('out-size'):
+                    out_size = pair_node.in_port(0).get_source().node.in_port(0).get_source().node['out-size']
+                    node.out_port(0).data.set_shape(np.array([1, out_size]))
+                elif pair_node.in_port(0).get_source().node.in_port(1).get_source().node.has_valid('out-size'):
+                    out_size = pair_node.in_port(0).get_source().node.in_port(1).get_source().node['out-size']
+                    node.out_port(0).data.set_shape(np.array([1, out_size]))
+            elif pair_node.in_port(0).get_source().node.op == "Mul" and \
+                    pair_node.in_port(0).get_source().node.in_port(0).get_source().node.has_valid('shape'):
+                out_size = pair_node.in_port(0).get_source().node.in_port(0).get_source().node.shape
```
Comment (pavel-esir) [inline model-optimizer/mo/ops/memoryoffset.py:72]: It looks ugly but it works

File: model-optimizer/mo/front/kaldi/extractors/tdnncomponent_ext.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        collect_until_token(pb, b'<TimeOffsets>')
+        time_offsets = read_binary_vector(pb, False, np.int32)
+
+        collect_until_token(pb, b'<LinearParams>')
+        weights, weights_shape = read_binary_matrix(pb)
+        collect_until_token(pb, b'<BiasParams>')
+        bias_params = read_binary_vector(pb)
+
+        collect_until_token(pb, b'<OrthonormalConstraint>')
+        orthonormal_constraint = read_binary_float_token(pb)  # used only on training
```
Comment (pavel-esir) [inline model-optimizer/mo/front/kaldi/extractors/tdnncomponent_ext.py:49]: this variables are not used but leaved them as 'prompt help' when parsing a Kaldi model.

File: model-optimizer/extensions/ops/MatMul.py [hunk 3]
```diff
@@ -221,7 +221,7 @@ def __init__(self, graph: Graph, attrs: dict):
 
 
 class FullyConnected(Op):
-    # TODO: remove `infer`, `type` and supported_attrs after op removal from IR Spec
+    # TODO: remove `infer` and `type` after op removal from IR Spec
```
Comment (pavel-esir) [inline model-optimizer/extensions/ops/MatMul.py:224]: I would remove `infer` and `type` but when I do that conversion fails.

File: model-optimizer/mo/front/kaldi/loader/utils.py [hunk 4]
```diff
@@ -61,6 +62,8 @@
     'splicecomponent',
     'sumgroupcomponent',
     'tanhcomponent',
+    'tdnncomponent',
+    'generaldropoutcomponent'
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/utils.py:66]: Add new entries in sorted order

File: model-optimizer/mo/middle/passes/infer.py [hunk 5]
```diff
@@ -108,7 +108,7 @@ def partial_infer(graph: Graph, start_node: str = None):
     graph.strict_mode = True
 
     # Mark all nodes as not inferred yet
-    if not start_node is None:
+    if start_node is not None:
```
Comment (lazarevevgeny) [inline model-optimizer/mo/middle/passes/infer.py:111]: Why did you change this?

File: model-optimizer/mo/ops/crop.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def _two_inputs_infer(node: Node):
         node['dim'] = dim
         node.out_node().shape = new_shape
 
-        if node.in_node(0).has_valid('value') and not node.graph.graph['cmd_params'].enable_ssd_gluoncv:
+        if node.in_node(0).has_valid('value') and not getattr(node.graph.graph['cmd_params'], 'enable_ssd_gluoncv', False):
```
Comment (lazarevevgeny) [inline model-optimizer/mo/ops/crop.py:167]: Why did you change this? Is the current logic broken?

File: model-optimizer/mo/ops/memoryoffset.py (hunk 1 above)
Comment (lazarevevgeny) [reply model-optimizer/mo/ops/memoryoffset.py:72]: Definitely need some comments here describing what is going on.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        return dict(
+            nodes=[
+                ('tdnncomponent', dict(kind='op', op='tdnncomponent'))
+            ],
+            edges=[]
+        )
+
+    @staticmethod
+    def replace_pattern(graph: Graph, match: dict):
+        tdnn_node: Node = match['tdnncomponent']
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:68]: Do not use pattern. Instead use something like this: `for tdnn_node in graph.get_op_nodes(op='tdnncomponent'):`

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/mo/ops/memoryoffset.py (hunk 1 above)
Comment (lazarevevgeny) [reply model-optimizer/mo/ops/memoryoffset.py:72]: Let's save `pair_node.in_port(0).get_source().node` to a dedicated variable with a meaningful name. It will make a code much more readable. Also, we may introduce additional variables and refactor the whole `else` block starting from line 50

Comment (pavel-esir) [reply model-optimizer/mo/ops/memoryoffset.py:72]: Agree, I was doing this at the moment

File: model-optimizer/mo/ops/crop.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/crop.py:167]: I don't know why but on Windows I had an error that namespace does not contain attribute `enable_ssd_gluoncv` while on Linux everything was ok, with the same commit. So I added this changes to get the same behavior.

File: model-optimizer/extensions/ops/MatMul.py [hunk 8]
```diff
@@ -234,11 +234,6 @@ def __init__(self, graph: Graph, attrs: dict):
             'out_ports_count': 1,
         }, attrs)
 
-    def supported_attrs(self):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/ops/MatMul.py]: done.

File: model-optimizer/extensions/front/kaldi/trainnodes_eraser.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+    @staticmethod
+    def pattern():
+        return dict(
+            nodes=[('identity', dict(kind='op', op='Identity'))],
+            edges=[]
+        )
+
+    @staticmethod
+    def replace_sub_graph(graph: Graph, match: dict):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/trainnodes_eraser.py]: Currently removed this. But let me explain why I had added it. Without this 'eraser' there was errors with converting a couple of models with train nodes where they are trickily connected to `MemoryOffset` nodes. Now I radically changed infer function of `MemoryOffset` and conversion works without this 'erases' on front phase.

File: model-optimizer/mo/front/kaldi/loader/utils.py (hunk 4 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/utils.py]: Done.

File: model-optimizer/mo/middle/passes/infer.py (hunk 5 above)
Comment (pavel-esir) [reply model-optimizer/mo/middle/passes/infer.py:111]: This logical expression was in `Joda style`. The last version is easier to understand, in my opinion. How did I get here? I was playing around this file because trying to fix/debug infer of `MemoryOffet` nodes.

File: model-optimizer/mo/ops/crop.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/crop.py:167]: Rebased to the latest master, still getting error on Win machine without there changes. While on linux -- everything is ok. @yekruglov, what you think about this?

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 7 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py]: Done. Please review

**Comment by pavel-esir** [comment]
@sadolini, @lazarevevgeny, I radically changed infer function of `MemoryOffset` op. Moved splitting `MemoryOffsets` to the `middle` phase and hence there is no need to lookup shapes of pair nodes (as it was done previously). So I leaved just one line instead of endless `ifs`. tdnn-f model at the moment do not match but model from #27253 match. Please review.

File: model-optimizer/extensions/ops/MatMul.py (hunk 3 above)
Comment (pavel-esir) [reply model-optimizer/extensions/ops/MatMul.py]: reverted changes.

**Comment by lazarevevgeny** [comment]
> @sadolini, @lazarevevgeny, I radically changed infer function of `MemoryOffset` op. Moved splitting `MemoryOffsets` to the `middle` phase and hence there is no need to lookup shapes of pair nodes (as it was done previously). So I leaved just one line instead of endless `ifs`. tdnn-f model at the moment do not match but model from #27253 match. Please review.

I think that this change resulted in a loop of operation dependencies. That is why the pre-commit failed.

**Comment by pavel-esir** [comment]
After minor corrections inference have matched. @sadolini @lazarevevgeny @jane-intel please review.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+from mo.front.common.replacement import FrontReplacementSubgraph
+from mo.graph.graph import Graph
+from mo.ops.memoryoffset import MemoryOffset
+from mo.ops.result import Result
+from mo.utils.error import Error
+
+
+class SplitLstmMemoryOffset(FrontReplacementSubgraph):
+    """
+    Splits MemoryOfffsets in LSTM blocks to avoid cycles during shape inference.
```
Comment (aalborov) [inline model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:27]: Extra `f`

**Review by aalborov — APPROVED** [review]

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py (hunk 10 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py]: Removed extra f and added more informative description of this transformation.

File: model-optimizer/mo/ops/crop.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/crop.py:167]: Communicated personally with @yekruglov. On windows I was calling MO by `mo_kaldi.py` so and in that case this variable is not initialized in `cmd_params` namespace. Therefore leaved this safety check for the sake of robustness.

File: model-optimizer/mo/ops/tdnncomponent.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+class TdnnComponent(Op):
+    op = 'tdnncomponent'
+    enabled = True
+
+    def __init__(self, graph: Graph, attrs: dict):
+        super().__init__(graph, {
+            'type': None,
+            'op': self.op,
+            'has_default': False,
```
Comment (sadolini) [inline model-optimizer/mo/ops/tdnncomponent.py:32]: Do you really need 'has_default' for TDNNComponent?

**Review by sadolini — CHANGES_REQUESTED** [review]

File: model-optimizer/mo/ops/tdnncomponent.py (hunk 11 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/tdnncomponent.py]: I copypasted code from memoryoffset and 'has_default' has left from there. Removed it.

File: model-optimizer/mo/front/kaldi/extractors/generaldropout_ext.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+ limitations under the License.
+"""
+
+from mo.front.extractor import FrontExtractorOp
+from mo.front.kaldi.loader.utils import read_binary_integer32_token, collect_until_token, \
+    read_binary_float_token
+from extensions.ops.identity import Identity
+
+
+class BatchNormComponentFrontExtractor(FrontExtractorOp):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/extractors/generaldropout_ext.py]: Agree. Corrected name.

**Review by sadolini — APPROVED** [review]

**Comment by jane-intel** [comment]
Merging according to @sadolini approval

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+                                           'element_size': offset_node['element_size'],
+                                           't': offset_node.t,
+                                           'has_default': offset_node.has_default}).create_node()
+        offset_node['splitted'] = True
+        offset_node.out_port(0).get_connection().set_source(paired_node.out_port(0))
+        res_node = Result(offset_node.graph, {'name': offset_node.id + "_output"}).create_node()
+        offset_node.out_port(0).connect(res_node.in_port(0))
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for offset_node in graph.get_op_nodes(op='MemoryOffset'):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:52]: nit: Should we add a check `not offset_node.soft_get('splitted', False)` before applying the `split_offset` function? I assume that when there is no cycle then the node is split already, but double check is probably a good idea.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+from mo.front.common.replacement import FrontReplacementSubgraph
+from mo.graph.graph import Graph
+from mo.ops.memoryoffset import MemoryOffset
+from mo.ops.result import Result
+from mo.utils.error import Error
+
+
+class SplitLstmMemoryOffset(FrontReplacementSubgraph):
+    """
+    Splits MemoryOffsets in LSTM blocks into 2 parts. These parts then will be converted to ReadValue and Assign.
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:27]: Why we say that this is an LSTM block? Could it be another type of recurrent network? Probably we should use word "recurrent" here...

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+          \_____________ _____|______________/____________/
+                           Concat
+                             |
+                        FullyConnected
+                             |
+    '''
+    enabled = True
+    run_not_recursively = True
+
+    def run_after(self):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:51]: Do we really need this dependency?

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+class SplitTdnnMemoryOffset(MiddleReplacementPattern):
+    '''
+    Splits MemoryOffsets in TDNN blocks into 2 parts. These parts then will be converted to ReadValue and Assign.
+    '''
+    enabled = True
+    run_not_recursively = True
+
+    def run_after(self):
+        from extensions.middle.PartialInfer import PartialInfer
+        return [PartialInfer]
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:32]: No need this dependency.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        for node in graph.get_op_nodes(op='tdnncomponent'):
+            self.replace_tdnn(graph, node)
+
+    def replace_tdnn(self, graph: Graph, tdnn_node: Node):
+        tdnn_name = tdnn_node.soft_get('name', tdnn_node.id)
+
+        concat_node = Concat(graph, {'axis': 1}).create_node()
+        rename_nodes([(tdnn_node, tdnn_name + '/to_be_removed'), (concat_node, tdnn_name)])
+
+        for i, t in enumerate(tdnn_node['time_offsets']):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:69]: nit: rename `i` to something more meaningful like `offset_ind`

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+            concat_node.add_input_port(i)
+            if t != 0:
+                memory_name = tdnn_name + '_memoryoffset_' + str(abs(t))
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:72]: nit: `_memoryoffset_` -> `/memoryoffset/`

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+                tdnn_node.in_port(0).get_source().connect(memoryoffset_node.in_port(0))
+                memoryoffset_node.out_port(0).connect(concat_node.in_port(i))
+            else:
+                # 0 time delay is not allowed in IE, it's meaningless
+                # if time offset is 0 then connect input of tdnncomponent directly to Concat without memoryoffset
+                tdnn_node.in_port(0).get_source().connect(concat_node.in_port(i))
+
+        weights = tdnn_node['weights']
+        bias_term = False
+        if tdnn_node.has_valid('biases') and tdnn_node['biases'] is not None:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:86]: No need to check `and tdnn_node['biases'] is not None` since `tdnn_node.has_valid('biases')` perform this check already.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+        fc_node = FullyConnected(graph, {'name': tdnn_name + '_fc',
+                                         'out-size': weights.shape[0],
+                                         'transpose_weights': True,
+                                         'bias_term': bias_term}).create_node()
+        concat_node.out_port(0).connect(fc_node.in_port(0))
+        tdnn_node.in_port(0).disconnect()
+        tdnn_node.out_port(0).get_connection().set_source(fc_node.out_port(0))
+
+        weights_node = Const(graph, {'name': tdnn_name + '_fc_weights', 'value': tdnn_node['weights']}).create_node()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:99]: nit: `_fc_weights` -> `/fc_weights`. The same comment for other places

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+                tdnn_node.in_port(0).get_source().connect(concat_node.in_port(i))
+
+        weights = tdnn_node['weights']
+        bias_term = False
+        if tdnn_node.has_valid('biases') and tdnn_node['biases'] is not None:
+            bias_term = True
+            biases = tdnn_node['biases']
+            assert len(biases) == weights.shape[0]
+
+        fc_node = FullyConnected(graph, {'name': tdnn_name + '_fc',
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:91]: Use `create_op_with_const_inputs` method to create `fc_node`.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+
+    def run_before(self):
+        from extensions.middle.ReplaceMemoryOffsetWithSplice import ReplaceMemoryOffsetWithMemoryNodePattern, ReplaceMemoryOffsetNodePattern
+        return [ReplaceMemoryOffsetNodePattern, ReplaceMemoryOffsetWithMemoryNodePattern]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for offset_node in graph.get_op_nodes(op='MemoryOffset', splitted=False):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:39]: What if `splitted` is `None`? Should the transformation be applied to ir?

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+            paired_node = MemoryOffset(graph, {'name': offset_node.pair_name, 'splitted': True, 'pair_name': offset_node.id,
+                                               't': offset_node.t, 'has_default': offset_node.has_default}).create_node()
+            offset_node['splitted'] = True
+            offset_node.out_port(0).get_connection().set_source(paired_node.out_port(0))
+            res_node = Result(graph, {'name': offset_node.id + "_output"}).create_node()
+            offset_node.out_port(0).connect(res_node.in_port(0))
+            paired_node['need_shape_inference'] = False
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:46]: Why do you disable shape inference?

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 24]
```diff
@@ -497,3 +498,25 @@ def parse_specifier(string, graph, layer_node_map):
...
+        scale_value = float(args[0])
+        layer_name = 'Scale_{}_{}'.format(node_name, scale_value)
+
+        if layer_name not in layer_node_map:
+            scale_name = graph.unique_id(prefix=layer_name)
+            scale_node = Mul(graph, {'name': scale_name}).create_node()
+
+            layer_node_map[layer_name] = scale_name
+
+            scale_const_name = 'Scale_{}_const_{}'.format(node_name, scale_value)
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:512]: Please, use the following naming convention for the node names `<prefix_node_name>/<additional_name>` everywhere. So here it should be `'{}/Scale/{}'.format(node_name, scale_value)`. 

BTW, do we really need the scale value to be the part of the node name?

File: model-optimizer/mo/ops/tdnncomponent.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+from mo.front.common.partial_infer.elemental import copy_shape_infer
+from mo.graph.graph import Graph, Node
+from mo.ops.op import Op
+from mo.utils.error import Error
+
+
+class TdnnComponent(Op):
+    op = 'tdnncomponent'
+    enabled = True
```
Comment (lazarevevgeny) [inline model-optimizer/mo/ops/tdnncomponent.py:26]: `enabled = True` -> `enabled = False` because the operation should not work as extractor.

File: model-optimizer/mo/ops/tdnncomponent.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+import numpy as np
+
+from mo.front.common.partial_infer.elemental import copy_shape_infer
+from mo.graph.graph import Graph, Node
+from mo.ops.op import Op
+from mo.utils.error import Error
```
Comment (lazarevevgeny) [inline model-optimizer/mo/ops/tdnncomponent.py:21]: Please, cleanup imports.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+    Splitting complicates shape inference but MemoryOffsets in LSTM blocks are cycled and, in order to make topological
+    sort possible during shape inference, they are splitted earlier on the front phase. In contrast, MemoryOffsets in
+    TDNN blocks are not cycled, so they will be splitted after shape infer on the middle.
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['fw'] == 'kaldi']
+
+    def run_after(self):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:36]: I don't remember why I putted this condition but now it works without it. Removed it.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+        from extensions.front.restore_ports import RestorePorts
+        return [RestorePorts]
+
+    def split_offset(self, offset_node):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:40]: Changed to static.

[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:40]: Agreed,

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+        offset_node.out_port(0).get_connection().set_source(paired_node.out_port(0))
+        res_node = Result(offset_node.graph, {'name': offset_node.id + "_output"}).create_node()
+        offset_node.out_port(0).connect(res_node.in_port(0))
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for offset_node in graph.get_op_nodes(op='MemoryOffset'):
+            try:
+                # if graph contains LSTM block with cycle, split MemoryOffset to enable shape infer
+                nx.find_cycle(graph, offset_node.id)
+                assert offset_node.has_valid('element_size')
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:56]: Moved.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py (hunk 13 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:52]: Before this transformation nowhere nodes are splitted and this `'splitted'` attribute is not changed to `True`.
> I assume that when there is no cycle then the node is split already
This statement is not always true. There can be cases when there is no cycle but not is not splitted yet. I advocate to leave this line without additional check.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 15 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:51]: Removed.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 16 above)
Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:32]: I checked. We don't need it. Removed.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 17 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:69]: Done.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 18 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:72]: Renamed.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 19 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:86]: Agreed.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 20 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:99]: Agreed. Renamed.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py (hunk 21 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py:91]: Agree. Changed.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 23 above)
Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:46]: If we do not set this attribute to `False` then after transformation it will try to `PartialInfer` and will fail because splitted node does not have `element_size` attribute. We either must (I) declare it here or (II) disable `PartialInfer` here. The first option seems redundant because data nodes on out_port 0 already have adequate shapes so I disabled `PartialInfer` for this node.

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 24 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py:512]: Agreed. `Scale` is redundant in the name. Removed.

File: model-optimizer/mo/ops/tdnncomponent.py (hunk 25 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/tdnncomponent.py:26]: Agreed. Thank for noting this.

File: model-optimizer/mo/ops/tdnncomponent.py (hunk 26 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/tdnncomponent.py:21]: Done.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 22 above)
Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:39]: `splitted` can not be None. At least is should newer be None. Currently there is no case when it can reach to value `None`. I don't think that we need some additional check for this case.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py (hunk 14 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:28]: You are right, in the general case there also can be other "recurrent" blocks not only LSTM and this transformation will also trigger it that cases. For the general case yes it will be reasonable to name `recurrent` instead of `LSTM` but currently the only recurrent block we have in Kaldi models is LSTM. To minimize efforts I prefer to leave the name as it is now. 

Do you insist to rename It?

**Comment by pavel-esir** [comment]
Replied to all comments. @lazarevevgeny, please review.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 22 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:35]: If the attribute is not set then it will be basically `None`. And it is not set then it should be considered as `False`. Am I right?

If so then we need this check. Or, you can add a default value for the `splitted` in the operation constructor. In this case you don't need to add any check here.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 23 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:42]: But what if sometime in the future someone will call shape inference for this node? It will fail and MO will crash. So let's set all necessary attributes so the shape infer will not fail.

File: model-optimizer/mo/ops/tdnncomponent.py (hunk 25 above)
Comment (lazarevevgeny) [reply model-optimizer/mo/ops/tdnncomponent.py:23]: Still not changed.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py (hunk 13 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:50]: Ok, agree. Added also `'splitted' = True` condition.

File: model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py (hunk 14 above)
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/split_lstm_memoryoffset.py:28]: Fair enough. Renamed LSTM -> recurrent. And mentioned in comments that at the moment we have only LSTM blocks with MemOffsets. Please review.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 22 above)
Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:35]: Agree. Added default value of `splitted = False`.

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py (hunk 23 above)
Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py]: Added setting of `element_size` here. Now if someone will call shape infer for this node it will not fail.

File: model-optimizer/mo/ops/tdnncomponent.py (hunk 25 above)
Comment (pavel-esir) [reply model-optimizer/mo/ops/tdnncomponent.py]: Sorry for that. Changed.

File: model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+        for node in graph.get_op_nodes(op='tdnncomponent'):
+            self.replace_tdnn(graph, node)
+
+    def replace_tdnn(self, graph: Graph, tdnn_node: Node):
+        tdnn_name = tdnn_node.soft_get('name', tdnn_node.id)
+
+        concat_node = Concat(graph, {'axis': 1}).create_node()
+        rename_nodes([(tdnn_node, tdnn_name + '/to_be_removed'), (concat_node, tdnn_name)])
+
+        for offfset_ind, t in enumerate(tdnn_node['time_offsets']):
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply model-optimizer/extensions/front/kaldi/tdnn_component_replacer.py]: done

**Comment by pavel-esir** [comment]
Agreed with all remarks, rebased from the latest master and resolved conflicts. @lazarevevgeny, please review

File: model-optimizer/extensions/middle/split_tdnn_memoryoffset.py [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+            offset_node.out_port(0).get_connection().set_source(paired_node.out_port(0))
+            res_node = Result(graph, {'name': offset_node.id + "_output"}).create_node()
+            offset_node.out_port(0).connect(res_node.in_port(0))
+
+            # If 'element_size' is previously copied from Parameter of from node with defined dim
+            if offset_node.has_valid('element_size'):
+                paired_node['element_size'] = offset_node['element_size']
+            # Copy shape from previous node. Typically (but not always) for TDNN blocks this is the case
+            else:
+                paired_node['element_size'] = offset_node.in_node().shape[1]
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/split_tdnn_memoryoffset.py:48]: `offset_node.in_node().shape[1]` -> `offset_node.in_port(0).data.get_shape()[1]`

Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py]: Done.

Comment (pavel-esir) [reply model-optimizer/extensions/middle/split_tdnn_memoryoffset.py]: Sorry for that. Now fixed it for real.

**Review by lazarevevgeny — APPROVED** [review]
